File: Python/String/12_word_wrap_probem.py
"""
Idea:
Using recursion (on 2 conditions-inserting words in same row OR in next row)
& using Matrix to store state of each word insertion.
"""

class Solution:
    def __init__(self):
        self.mat = [[-1] * 2001 for _ in range(501)]

    def ReccFunction(self, i, remainder, arr, k):
        if i == len(arr):  # Base case
            return 0

        if self.mat[i][remainder] != -1:  # If the case is already explored, return the value
            return self.mat[i][remainder]

        if arr[i] > remainder:
            # If the word can't be inserted into the same line, insert the word in the next line
            ans = (remainder + 1) * (remainder + 1) + self.ReccFunction(i + 1, k - arr[i] - 1, arr, k)
        else:
            # If we can insert the word in the same line, here are 2 cases
            choice1 = self.ReccFunction(i + 1, remainder - arr[i] - 1, arr, k)  # If the word is inserted in the same line
            choice2 = (remainder + 1) * (remainder + 1) + self.ReccFunction(i + 1, k - arr[i] - 1, arr, k)  # If the word is inserted in a new line
            ans = min(choice1, choice2)

        self.mat[i][remainder] = ans  # Store the calculated value for future use
        return self.mat[i][remainder]

    def solveWordWrap(self, nums, k):
        return self.ReccFunction(0, k, nums, k)

if __name__ == '__main__':
    T = int(input())
    for i in range(T):
        n = int(input())
        nums = list(map(int, input().split()))
        k = int(input())
        obj = Solution()
        ans = obj.solveWordWrap(nums, k)
        print(ans)

File: C++/String/12_word_wrap_probem.cpp
/*
Idea:
Using recursion (on 2 conditions-inserting words in same row OR in next row)
& using Matrix to store state of each word insertion.
*/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int mat[501][2001]; // martrix to store state values
    int ReccFunction(int i, int remainder, vector<int>& arr, int& k)
    {
        if (i == arr.size()) // base case
            return 0;

        if (mat[i][remainder] != -1) // if case is already explored, return the value
            return mat[i][remainder];

        int ans;
        if (arr[i] > remainder)
            // if word can't be inserted into same line, inserting word in next line 
            ans = (remainder + 1) * (remainder + 1) + ReccFunction(i + 1, k - arr[i] - 1, arr, k);
        else
        {
            // if we can insert word in same line, here are 2 cases
            int choice1 = ReccFunction(i + 1, remainder - arr[i] - 1, arr, k); // if word is inserted in same line
            int choice2 = (remainder + 1) * (remainder + 1) + ReccFunction(i + 1, k - arr[i] - 1, arr, k); // if word is inserted in new line
            ans = min(choice1, choice2);
        }
        mat[i][remainder] = ans; // storing value calculated for future use
        return mat[i][remainder];
    }
    int solveWordWrap(vector<int>nums, int k)
    {
        memset(mat, -1, sizeof(mat));
        return ReccFunction(0, k, nums, k);
    }
};

int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        int n, k;
        cin >> n;
        vector<int>nums(n);
        for (int i = 0; i < n; i++)cin >> nums[i];
        cin >> k;
        Solution obj;
        cout << obj.solveWordWrap(nums, k) << endl;
    }
    return 0;
}